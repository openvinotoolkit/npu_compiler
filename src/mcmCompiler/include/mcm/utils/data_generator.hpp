#ifndef DATA_GENERATOR_HPP_
#define DATA_GENERATOR_HPP_

#include <vector>
#include <cstdint>
#include <random>
#include <algorithm>
#include <fstream>

namespace mv
{

    namespace utils
    {

        template <class T_data>
        std::vector<T_data> generateSequence(std::size_t dataSize)
        {
            std::vector<T_data> result(dataSize);

            for (std::size_t i = 0; i < result.size(); ++i)
                result[i] = (T_data)i;

            return result;

        }

        template <class T_data>
        std::vector<T_data> generateSequence(std::size_t dataSize, T_data start, T_data dt)
        {
            std::vector<T_data> result(dataSize);

            T_data val = start;
            for (std::size_t i = 0; i < result.size(); ++i)
            {
                result[i] = val;
                val += dt;
            }

            return result;

        }

        // Seed is given for reproducibility of results
        template <class T_data>
        std::vector<T_data> generateRandomSequence(std::size_t dataSize, T_data start, T_data end, unsigned seed)
        {
            // Specify the engine and distribution.
            std::mt19937 mersenne_engine {seed};  // Generates random integers
            std::uniform_int_distribution<T_data> dist {start, end};

            auto gen = [&dist, &mersenne_engine](){
                           return dist(mersenne_engine);
                       };

            std::vector<T_data> result(dataSize);
            std::generate(result.begin(), result.end(), gen);

            return result;

        }

        template <class T_data>
        std::vector<T_data> readWeightsFromFile(const std::string& fileName)
        {
            std::ifstream is (fileName, std::ifstream::binary);

            is.seekg (0, is.end);
            std::size_t length = is.tellg();
            is.seekg (0, is.beg);
            char * buffer = new char [length];
            is.read (buffer,length);
            is.close();
            std::vector<T_data> result(length/sizeof(T_data));
            memcpy(result.data(), buffer, length);
            delete[] buffer;

            return result;
        }

    }

}

#endif // DATA_GENERATOR_HPP
