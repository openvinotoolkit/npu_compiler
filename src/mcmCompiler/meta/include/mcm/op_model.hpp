/*
    DO NOT MODIFY - that file was generated automatically using op::OpRegistry::generateCompositionAPI()
*/

#ifndef MV_OP_MODEL_HPP_
#define MV_OP_MODEL_HPP_

#include "include/mcm/compositional_model.hpp"
#include "include/mcm/computation/model/base_op_model.hpp"

#include "include/mcm/compiler/compilation_profiler.hpp"

namespace mv

{

    class OpModel: public BaseOpModel, public CompositionalModel
    {

    private:
    public:

        OpModel(const std::string& name);
        OpModel(ComputationModel& model);
        virtual ~OpModel();

        mv::Data::TensorIterator align(const std::string& name, Data::TensorIterator data, const std::size_t& dimension = 2, const std::size_t& pad = 16) override;
        mv::Data::TensorIterator argmax(const std::string& name, Data::TensorIterator data, const int64_t& out_max_val, const int64_t& top_k, const int64_t& axis = 99) override;
        mv::Data::TensorIterator averagePool(const std::string& name, Data::TensorIterator data, const std::array<unsigned short, 2>& kSize, const std::array<unsigned short, 2>& stride, const std::array<unsigned short, 4>& padding, const bool& exclude_pad = true) override;
        mv::Data::TensorIterator barrierTask(const std::string& name, const Barrier& Barrier);
        mv::Data::TensorIterator batchNormalization(const std::string& name, Data::TensorIterator data, Data::TensorIterator mean, Data::TensorIterator variance, Data::TensorIterator offset, Data::TensorIterator scale, const double& eps) override;
        mv::Data::TensorIterator bias(const std::string& name, Data::TensorIterator data, Data::TensorIterator weights) override;
        mv::Data::TensorIterator cTCDecoder(const std::string& name, Data::TensorIterator data, Data::TensorIterator seq, const bool& ctc_merge_repeated) override;
        mv::Data::TensorIterator concat(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::string& axis = "C") override;
        mv::Data::TensorIterator constant(const std::string& name, const std::vector<double>& data, const Shape& shape, const DType& dType, const Order& order) override;
        mv::Data::TensorIterator constantDataElement(const std::string& name, const std::vector<mv::DataElement>& data, const Shape& shape, const DType& dType, const Order& order);
        mv::Data::TensorIterator constantInt(const std::string& name, const std::vector<int64_t>& data, const Shape& shape, const DType& dType, const Order& order) override;
        mv::Data::TensorIterator conv(const std::string& name, Data::TensorIterator data, Data::TensorIterator weights, const std::array<unsigned short, 2>& stride, const std::array<unsigned short, 4>& padding, const unsigned& dilationFactor = 1, const unsigned& group = 1) override;
        mv::Data::TensorIterator conversion(const std::string& name, Data::TensorIterator data, const Order& order);
        mv::Data::TensorIterator copy(const std::string& name, Data::TensorIterator data) override;
        mv::Data::TensorIterator crop(const std::string& name, Data::TensorIterator data, const std::size_t& cropVal, const std::size_t& dimension = 2) override;
        mv::Data::TensorIterator custom(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::vector<uint8_t>& kernelData, const std::vector<uint8_t>& paramData, const std::vector<mv::TensorInfo>& outputsInfo) override;
        mv::Data::TensorIterator dMATask(const std::string& name, Data::TensorIterator data, const DmaDirection& direction, const uint8_t& port = 0);
        mv::Data::TensorIterator dPUTaskConv(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::array<unsigned short, 2>& stride, const std::array<unsigned short, 4>& padding, const unsigned& dilationFactor = 1, const unsigned& group = 1);
mv::Data::TensorIterator dPUTaskMaxPool(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::array<unsigned short, 2>& kSize, const std::array<unsigned short, 2>& stride, const std::array<unsigned short, 4>& padding, const bool& exclude_pad = true);
mv::Data::TensorIterator dPUTaskDepthwiseConv(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::array<unsigned short, 2>& stride, const std::array<unsigned short, 4>& padding, const unsigned& dilationFactor = 1);
mv::Data::TensorIterator dPUTaskEltwise(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::string& eltwiseType);
        mv::Data::TensorIterator deallocate(const std::string& name, Data::TensorIterator inputs);
        mv::Data::TensorIterator deconv(const std::string& name, Data::TensorIterator data, Data::TensorIterator weights, const std::array<unsigned short, 2>& stride, const std::array<unsigned short, 4>& padding, const unsigned& dilationFactor = 1, const unsigned& group = 1, const bool& is_depthwise = false) override;
        mv::Data::TensorIterator depthwiseConv(const std::string& name, Data::TensorIterator data, Data::TensorIterator weights, const std::array<unsigned short, 2>& stride, const std::array<unsigned short, 4>& padding, const unsigned& dilationFactor = 1) override;
        mv::Data::TensorIterator detectionOutput(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const int64_t& num_classes, const int64_t& keep_top_k, const double& nms_threshold, const int64_t& background_label_id, const int64_t& top_k, const bool& variance_encoded_in_target, const std::string& code_type, const bool& share_location, const double& confidence_threshold, const bool& clip_before_nms, const bool& clip_after_nms, const int64_t& decrease_label_id, const bool& normalized, const int64_t& input_height, const int64_t& input_width, const double& objectness_score) override;
        mv::Data::TensorIterator dropout(const std::string& name, Data::TensorIterator input) override;
        mv::Data::TensorIterator dummy(const std::string& name, Data::TensorIterator data) override;
        mv::Data::TensorIterator eltwise(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::string& eltwiseType) override;
        mv::Data::TensorIterator elu(const std::string& name, Data::TensorIterator data, const unsigned& alpha = 1) override;
        mv::Data::TensorIterator exp(const std::string& name, Data::TensorIterator data) override;
        mv::Data::TensorIterator fakeQuantize(const std::string& name, Data::TensorIterator data, Data::TensorIterator input_min, Data::TensorIterator input_max, Data::TensorIterator output_min, Data::TensorIterator output_max, const unsigned& levels) override;
        mv::Data::TensorIterator flatten(const std::string& name, Data::TensorIterator input, const int64_t& axis = 1, const int64_t& end_axis = 3) override;
        mv::Data::TensorIterator fullyConnected(const std::string& name, Data::TensorIterator data, Data::TensorIterator weights) override;
        mv::Data::TensorIterator gather(const std::string& name, Data::TensorIterator data, Data::TensorIterator indices, const unsigned& axis) override;
        mv::Data::TensorIterator hSwish(const std::string& name, Data::TensorIterator data) override;
        mv::Data::TensorIterator identity(const std::string& name, Data::TensorIterator data) override;
        mv::Data::TensorIterator implicitConcat(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::string& axis = "C");
        mv::Data::TensorIterator implicitInput(const std::string& name, Data::TensorIterator data, const Shape& shape, const DType& dType, const Order& order) override;
        mv::Data::TensorIterator implicitInputSlice(const std::string& name, Data::TensorIterator inputs);
        mv::Data::TensorIterator implicitJoin(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::string& axis = "HW");
        mv::Data::TensorIterator implicitOutput(const std::string& name, Data::TensorIterator data) override;
        mv::Data::TensorIterator implicitPermute(const std::string& name, Data::TensorIterator inputs, const Shape& shape);
        mv::Data::TensorIterator implicitReshape(const std::string& name, Data::TensorIterator inputs, const Shape& shape);
        mv::Data::TensorIterator implicitUnion(const std::string& name, const std::vector< Data::TensorIterator >& inputs);
        mv::Data::TensorIterator input(const std::string& name, const Shape& shape, const DType& dType, const Order& order, const bool& networkInput = true) override;
        mv::Data::TensorIterator interp(const std::string& name, Data::TensorIterator data, const double& factor, const unsigned& pad_beg, const unsigned& pad_end, const unsigned& height = 0, const unsigned& width = 0, const bool& align_corners = true) override;
        mv::Data::TensorIterator leakyRelu(const std::string& name, Data::TensorIterator data, const double& alpha = 0) override;
        mv::Data::TensorIterator localResponseNormalization(const std::string& name, Data::TensorIterator data, const unsigned& size, const unsigned& bias) override;
        mv::Data::TensorIterator matMul(const std::string& name, Data::TensorIterator data0, Data::TensorIterator data1) override;
        mv::Data::TensorIterator maxPool(const std::string& name, Data::TensorIterator data, const std::array<unsigned short, 2>& kSize, const std::array<unsigned short, 2>& stride, const std::array<unsigned short, 4>& padding, const bool& exclude_pad = true) override;
        mv::Data::TensorIterator maximum(const std::string& name, Data::TensorIterator inputs, const double& maximum) override;
        mv::Data::TensorIterator minimum(const std::string& name, Data::TensorIterator inputs, const double& minimum) override;
        mv::Data::TensorIterator norm(const std::string& name, Data::TensorIterator data, const double& alpha, const double& beta, const std::string& region, const unsigned& local_size) override;
        mv::Data::TensorIterator normalize(const std::string& name, Data::TensorIterator data, Data::TensorIterator weights, const double& eps, const unsigned& across_spatial = 0, const unsigned& channel_shared = 0) override;
        mv::Data::TensorIterator output(const std::string& name, Data::TensorIterator data, const DType& precision = mv::DType("Default"), const bool& networkOutput = true) override;
        mv::Data::TensorIterator pSROIPooling(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::size_t& output_dim, const std::size_t& group_size, const double& spatial_scale, const std::size_t& pooled_w, const std::size_t& pooled_h, const std::size_t& spatial_bin_x, const std::size_t& spatial_bin_y, const std::string& mode) override;
        mv::Data::TensorIterator permute(const std::string& name, Data::TensorIterator data, const Order& order) override;
        mv::Data::TensorIterator placeholderTask(const std::string& name, const Shape& shape, const DType& dType, const Order& order);
        mv::Data::TensorIterator prelu(const std::string& name, Data::TensorIterator data, Data::TensorIterator slope) override;
        mv::Data::TensorIterator priorbox(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const unsigned& flip, const unsigned& clip, const double& step_w, const double& step_h, const double& offset) override;
        mv::Data::TensorIterator proposal(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::vector<float>& scale, const std::vector<float>& ratio, const unsigned& base_size, const unsigned& pre_nms_topn, const unsigned& post_nms_topn, const double& nms_thresh, const unsigned& feat_stride, const unsigned& min_size, const double& pre_nms_thresh = 0.000000000000000, const bool& clip_before_nms = true, const bool& clip_after_nms = false, const bool& normalize = false, const double& box_size_scale = 1.000000000000000, const double& box_coordinate_scale = 1.000000000000000, const std::string& framework = "TENSORFLOW", const bool& for_deformable = false) override;
        mv::Data::TensorIterator pseudoOp(const std::string& name, const std::vector< Data::TensorIterator >& inputs);
        mv::Data::TensorIterator quantize(const std::string& name, Data::TensorIterator data) override;
        mv::Data::TensorIterator rOIPooling(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const unsigned& pooled_w, const unsigned& pooled_h, const double& spatial_scale, const unsigned& roi_pooling_method, const unsigned& num_rois) override;
        mv::Data::TensorIterator reciprocal(const std::string& name, Data::TensorIterator data) override;
        mv::Data::TensorIterator refConv(const std::string& name, Data::TensorIterator data, Data::TensorIterator weights, const std::array<unsigned short, 2>& stride, const std::array<unsigned short, 4>& padding, const unsigned& dilationFactor = 1, const unsigned& group = 1);
        mv::Data::TensorIterator regionYolo(const std::string& name, Data::TensorIterator data, const unsigned& coords, const unsigned& classes, const bool& do_softmax, const unsigned& num = 0, const std::vector<unsigned>& mask = {}) override;
        mv::Data::TensorIterator relu(const std::string& name, Data::TensorIterator data) override;
        mv::Data::TensorIterator reorder(const std::string& name, Data::TensorIterator data, const Order& order) override;
        mv::Data::TensorIterator reorgYolo(const std::string& name, Data::TensorIterator data, const unsigned& stride) override;
        mv::Data::TensorIterator resample(const std::string& name, Data::TensorIterator input, const std::string& interpolation, const bool& antialias, const Shape& output_shape) override;
        mv::Data::TensorIterator reshape(const std::string& name, Data::TensorIterator data, const Shape& shape) override;
        mv::Data::TensorIterator scale(const std::string& name, Data::TensorIterator data, Data::TensorIterator weights) override;
        mv::Data::TensorIterator sigmoid(const std::string& name, Data::TensorIterator data) override;
        mv::Data::TensorIterator slice(const std::string& name, Data::TensorIterator data, const Shape& begin, const Shape& size) override;
        mv::Data::TensorIterator softmax(const std::string& name, Data::TensorIterator data, const std::string& axis = "C") override;
        mv::Data::TensorIterator sparsityMap(const std::string& name, const std::vector<int64_t>& data, const Shape& shape, const DType& dType, const Order& order);
        mv::Data::TensorIterator tanh(const std::string& name, Data::TensorIterator data) override;
        mv::Data::TensorIterator tile(const std::string& name, Data::TensorIterator data, const unsigned& axis, const unsigned& tiles) override;
        mv::Data::TensorIterator topK(const std::string& name, Data::TensorIterator data, const std::string& sort, const std::string& mode, const int64_t& top_k, const int64_t& axis = 99) override;
        mv::Data::TensorIterator uPATaskDummy(const std::string& name, const std::vector< Data::TensorIterator >& inputs);
mv::Data::TensorIterator uPATaskIdentity(const std::string& name, const std::vector< Data::TensorIterator >& inputs);
mv::Data::TensorIterator uPATaskSoftmax(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::string& axis = "C");
mv::Data::TensorIterator uPATaskProposal(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::vector<float>& scale, const std::vector<float>& ratio, const unsigned& base_size, const unsigned& pre_nms_topn, const unsigned& post_nms_topn, const double& nms_thresh, const unsigned& feat_stride, const unsigned& min_size, const double& pre_nms_thresh = 0.000000000000000, const bool& clip_before_nms = true, const bool& clip_after_nms = false, const bool& normalize = false, const double& box_size_scale = 1.000000000000000, const double& box_coordinate_scale = 1.000000000000000, const std::string& framework = "TENSORFLOW", const bool& for_deformable = false);
mv::Data::TensorIterator uPATaskROIPooling(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const unsigned& pooled_w, const unsigned& pooled_h, const double& spatial_scale, const unsigned& roi_pooling_method, const unsigned& num_rois);
mv::Data::TensorIterator uPATaskPSROIPooling(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::size_t& output_dim, const std::size_t& group_size, const double& spatial_scale, const std::size_t& pooled_w, const std::size_t& pooled_h, const std::size_t& spatial_bin_x, const std::size_t& spatial_bin_y, const std::string& mode);
mv::Data::TensorIterator uPATaskQuantize(const std::string& name, const std::vector< Data::TensorIterator >& inputs);
mv::Data::TensorIterator uPATaskReshape(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const Shape& shape);
mv::Data::TensorIterator uPATaskRegionYolo(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const unsigned& coords, const unsigned& classes, const bool& do_softmax, const unsigned& num = 0, const std::vector<unsigned>& mask = {});
mv::Data::TensorIterator uPATaskReorgYolo(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const unsigned& stride);
mv::Data::TensorIterator uPATaskNormalize(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const double& eps, const unsigned& across_spatial = 0, const unsigned& channel_shared = 0);
mv::Data::TensorIterator uPATaskPermute(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const Order& order);
mv::Data::TensorIterator uPATaskEltwise(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::string& eltwiseType);
mv::Data::TensorIterator uPATaskInterp(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const double& factor, const unsigned& pad_beg, const unsigned& pad_end, const unsigned& height = 0, const unsigned& width = 0, const bool& align_corners = true);
mv::Data::TensorIterator uPATaskDetectionOutput(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const int64_t& num_classes, const int64_t& keep_top_k, const double& nms_threshold, const int64_t& background_label_id, const int64_t& top_k, const bool& variance_encoded_in_target, const std::string& code_type, const bool& share_location, const double& confidence_threshold, const bool& clip_before_nms, const bool& clip_after_nms, const int64_t& decrease_label_id, const bool& normalized, const int64_t& input_height, const int64_t& input_width, const double& objectness_score);
mv::Data::TensorIterator uPATaskPriorbox(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const unsigned& flip, const unsigned& clip, const double& step_w, const double& step_h, const double& offset);
mv::Data::TensorIterator uPATaskArgmax(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const int64_t& out_max_val, const int64_t& top_k, const int64_t& axis = 99);
mv::Data::TensorIterator uPATaskTopK(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::string& sort, const std::string& mode, const int64_t& top_k, const int64_t& axis = 99);
mv::Data::TensorIterator uPATaskNorm(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const double& alpha, const double& beta, const std::string& region, const unsigned& local_size);
mv::Data::TensorIterator uPATaskResample(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::string& interpolation, const bool& antialias, const Shape& output_shape);
mv::Data::TensorIterator uPATaskFakeQuantize(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const unsigned& levels);
mv::Data::TensorIterator uPATaskCustom(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::vector<uint8_t>& kernelData, const std::vector<uint8_t>& paramData, const std::vector<mv::TensorInfo>& outputsInfo);
mv::Data::TensorIterator uPATaskSigmoid(const std::string& name, const std::vector< Data::TensorIterator >& inputs);
mv::Data::TensorIterator uPATaskDeconv(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::array<unsigned short, 2>& stride, const std::array<unsigned short, 4>& padding, const unsigned& dilationFactor = 1, const unsigned& group = 1, const bool& is_depthwise = false);
mv::Data::TensorIterator uPATaskTile(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const unsigned& axis, const unsigned& tiles);
mv::Data::TensorIterator uPATaskCTCDecoder(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const bool& ctc_merge_repeated);
mv::Data::TensorIterator uPATaskRefConv(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const std::array<unsigned short, 2>& stride, const std::array<unsigned short, 4>& padding, const unsigned& dilationFactor = 1, const unsigned& group = 1);
mv::Data::TensorIterator uPATaskGather(const std::string& name, const std::vector< Data::TensorIterator >& inputs, const unsigned& axis);
mv::Data::TensorIterator uPATaskHSwish(const std::string& name, const std::vector< Data::TensorIterator >& inputs);
        mv::Data::TensorIterator weightsTable(const std::string& name, const std::vector<int64_t>& data, const Shape& shape, const DType& dType, const Order& order);

        Data::OpListIterator getSourceOp(Data::TensorIterator tensor) override;
        void addAttr(Data::OpListIterator op, const std::string& name, const Attribute& attr) override;
        bool isValid() const override;
        bool isValid(Data::TensorIterator tensor) const override;
        bool isValid(Data::OpListIterator op) const override;
        std::string getName() const override;

    };

}

#endif //MV_OP_MODEL_HPP_
