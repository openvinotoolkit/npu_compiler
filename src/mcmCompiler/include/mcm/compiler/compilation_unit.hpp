#ifndef MV_COMPILATION_UNIT_HPP_
#define MV_COMPILATION_UNIT_HPP_

#include "include/mcm/computation/model/control_model.hpp"
#include "include/mcm/computation/model/data_model.hpp"
#include "include/mcm/op_model.hpp"
#include "include/mcm/logger/log_sender.hpp"
#include "include/mcm/pass/pass_manager.hpp"
#include "include/mcm/utils/env_loader.hpp"
#include "include/mcm/compiler/compilation_descriptor.hpp"
#include "include/mcm/compiler/compilation_profiler.hpp"
#include "include/mcm/target/kmb/runtime_model/runtime_model.hpp"

namespace mv
{

    class CompilationUnit : public LogSender
    {

        static const std::string ma2490DefTargetDescPath_;
        static const std::string ma3100DefTargetDescPath_;
        static const std::string ma3600DefTargetDescPath_;

        static const std::string ma2490DefCompDescPath_;
        static const std::string ma3100DefCompDescPath_;
        static const std::string ma3600DefCompDescPath_;
        static const std::string compositionalModelRecordingsPath_;
        static const unsigned jsonParserBufferLength_ = 256;

        static Logger& logger_;
        static bool init_;

        OpModel* model_;
        PassManager passManager_;
        TargetDescriptor targetDescriptor_;
        CompilationDescriptor compDescriptor_;
        bool preCompiled_;

    public:

        CompilationUnit(const std::string& modelName);
        CompilationUnit(const char *blobBuffer, unsigned blobSize, TargetDescriptor td);
        CompilationUnit(const CompilationUnit & rhs) = delete;
        ~CompilationUnit();

        CompilationUnit & operator=(const CompilationUnit & rhs) = delete;
        bool loadTargetDescriptor(const std::string& path);
        bool loadTargetDescriptor(Target target);
        bool loadCompilationDescriptor(const std::string& path);
        bool loadCompilationDescriptor(Target target);

        CompilationDescriptor& compilationDescriptor();
        OpModel& model();

        bool initialize();
        Element runStep();
        Element run();
        bool completed() const;
        void reset();

        void getName(char* name, unsigned bufferSize) const;
        std::shared_ptr<std::vector<char> > getBlob() const;
        const BufferMap& getBufferMap() const;

        void loadBlob(const char *blobBuffer, unsigned blobSize);

        virtual std::string getLogID() const override;

    };

}

#endif // MV_COMPILATION_UNIT_HPP_
