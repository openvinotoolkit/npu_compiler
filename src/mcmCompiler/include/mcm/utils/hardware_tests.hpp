//
// Copyright (C) 2022 Intel Corporation
// SPDX-License-Identifier: Apache 2.0
//
#ifndef MCM_HARDWARE_TESTS_
#define MCM_HARDWARE_TESTS_

#include "include/mcm/compiler/compilation_unit.hpp"
#include <string>
#include <iostream>

namespace mv
{
    struct ReturnCodes
    {
        int mcmBlobOnHardware;
        int fathomCompilation;
        int diffOutput;
        int fathomVsCaffe;
        int fathomVsMcm;
    };

    //NOTE: To execute these tests GenerateCaffe pass must be enable in adaptation passes.
    ReturnCodes HWTest(mv::CompilationUnit& unit, std::string outputName, bool fathomHardware = false);
    void printReport(mv::ReturnCodes returnValue, std::ostream& out);

}


#endif
