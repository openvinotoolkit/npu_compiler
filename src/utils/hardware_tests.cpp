#include "include/mcm/utils/hardware_tests.hpp"

void mv::HWTest(mv::CompilationUnit& unit, std::string outputName)
{
    auto compOutput = unit.run();

    system(std::string("dot -Tsvg " + outputName + ".dot -o " + outputName + ".svg").c_str());
    system(std::string("dot -Tsvg " + outputName + "_adapt.dot -o " + outputName + "_adapt.svg").c_str());
    system(std::string("dot -Tsvg " + outputName + "_final.dot -o " + outputName + "_final.svg").c_str());

    //0) Run generated blob
    std::cout << "RUNNING ON HW MCMCOMPILER BLOB" << std::endl;
    std::string command0("python3 " + mv::utils::projectRootPath() + "/python/tools/mcmRunHW.py --blob " + outputName + ".blob --result " + outputName + ".npy --image " + mv::utils::projectRootPath() + "/mug.png");
    std::cout << command0 << std::endl;
    system(command0.c_str());

    //1) Compile generated prototxt with fathom
    std::cout << "COMPILING GENERATED PROTOTXT WITH FATHOM" << std::endl;
    std::string command1("python3 " + mv::utils::mdkRootPath() + "/projects/Fathom/src2/mvNCCompile.py " + outputName + ".prototxt -w " + outputName + ".caffemodel ");
    std::cout << command1 << std::endl;
    system(command1.c_str());

    //2) Compare python and caffe
    std::cout << "COMPARING PYTHON AND CAFFE" << std::endl;
    std::string command3("python3 " + mv::utils::mdkRootPath() + "/projects/Fathom/src2/mvNCCheck.py " + outputName + ".prototxt -w " + outputName + ".caffemodel ");
    std::cout << command3 << std::endl;
    system(command3.c_str());

    //3) Compare python and cpp
    std::cout << "COMPARING FATHOM AND MCMCOMPILER BLOBS" << std::endl;
    std::string command2(mv::utils::projectRootPath() + "/python/tools/mcmCheck.sh -b " + outputName + ".blob -b graph -i " + mv::utils::projectRootPath() + "/mug.png");
    std::cout << command2 << std::endl;
    system(command2.c_str());


}
