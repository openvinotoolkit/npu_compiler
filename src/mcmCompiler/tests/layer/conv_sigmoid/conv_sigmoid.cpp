//This file is the parsed network which is created through python.
#include "include/mcm/compiler/compilation_unit.hpp"
#include "include/mcm/utils/data_generator.hpp"
#include "include/mcm/op_model.hpp"
#include "include/mcm/utils/hardware_tests.hpp"

#include <iostream>
#include <fstream>

int main()
{
    mv::CompilationUnit unit("ConvSigmoid");
    mv::OpModel& om = unit.model();

    auto input_shape = mv::Shape({1,1,100,1});
    auto weights_shape = mv::Shape({1,1,100,1000});

    // Load weights (u8 saved as int64)
    std::string weights_filename(mv::utils::projectRootPath() + "/tests/layer/conv_sigmoid/conv_sigmoid.w");
    std::vector<int64_t> weightsData0 = mv::utils::readWeightsFromFile<int64_t>(weights_filename);

    // Load biases (int32 saved as int64)
    std::string bias_filename(mv::utils::projectRootPath() + "/tests/layer/conv_sigmoid/conv_sigmoid.b");
    std::vector<int64_t> biasWeightsData0 = mv::utils::readWeightsFromFile<int64_t>(bias_filename);

    // Input
    auto input0 = om.input({1,1,184,1}, mv::DType("UInt8"), mv::Order::getZMajorID(4), {{128},{0.007843137718737125},{-1.0},{1.0}}, "input#20");

    // FC
    auto weights0 = om.constantInt(weightsData0,{184,128}, mv::DType("UInt8"), mv::Order("WC"), {{128},{0.001087399199604988},{-0.13866370916366577},{0.13862307369709015}}, "fully_connected/BiasAdd_weights#1");
    auto fc0 = om.fullyConnected(input0, weights0, mv::DType("UInt8"), {{128},{0.0470588244497776},{-6.023529529571533},{5.976470470428467}}, "fully_connected/BiasAdd#21");
    auto biasWeights0 = om.constantInt(biasWeightsData0,{128}, mv::DType("UInt8"), mv::Order::getColMajorID(1), {{0},{8.528621037839912e-06},{},{}}, "fully_connected/BiasAdd_bias#2");
    auto bias_i0 = om.bias(fc0, biasWeights0, mv::DType("UInt8"), {{128},{0.0470588244497776},{-6.023529529571533},{5.976470470428467}});

    // Sigmoid
    auto sigmoid0 = om.sigmoid(bias_i0, mv::DType("UInt8"), {{0},{0.00390625},{0.0},{0.99609375}}, "fully_connected/Sigmoid#22");

    // Output
    om.output(sigmoid0);

    std::string compDescPath = mv::utils::projectRootPath() + "/config/compilation/release_kmb.json";
    unit.loadCompilationDescriptor(compDescPath);
    unit.loadTargetDescriptor(mv::Target::ma2490);
    unit.initialize();
    unit.run();

    return 0;
}
