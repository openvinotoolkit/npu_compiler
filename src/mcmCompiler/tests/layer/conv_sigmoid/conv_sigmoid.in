version https://git-lfs.github.com/spec/v1
oid sha256:dcccd2d5c1d8e33c416777eea613e434a5af29632fda837a17eb3a5d813259bf
size 184
