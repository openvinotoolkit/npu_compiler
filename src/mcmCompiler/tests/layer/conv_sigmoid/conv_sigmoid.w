version https://git-lfs.github.com/spec/v1
oid sha256:d900f372f6541bde1a74e042f7b52adf608b4817fbe0f7a5f47f7360ad9e1da9
size 188416
