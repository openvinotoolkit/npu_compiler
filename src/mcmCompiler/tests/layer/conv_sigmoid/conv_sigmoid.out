version https://git-lfs.github.com/spec/v1
oid sha256:f5a072eeefc2fe28d86fb389fd2f0eb1ec35a5dbec78e2a888c4ef26e01d1018
size 128
